runtime/v2/runc/task/service.go

```diff
@@ -137,11 +137,14 @@ type containerProcess struct {
 // that its exit can be handled efficiently. If the process has already exited,
 // it handles the exit immediately. handleStarted should be called after the
 // event announcing the start of the process has been published.
+// Note that handleStarted needs to be aware of whether s.mu is already held
+// when it is called. If s.mu has been held, handleProcessExitNoLock should be
+// used to avoid deadlock.
 //
 // The returned cleanup closure releases resources used to handle early exits.
 // It must be called before the caller of preStart returns, otherwise severe
 // memory leaks will occur.
-func (s *service) preStart(c *runc.Container) (handleStarted func(*runc.Container, process.Process), cleanup func()) {
+func (s *service) preStart(c *runc.Container, muLocked bool) (handleStarted func(*runc.Container, process.Process), cleanup func()) {
```

Comment:
`preStart` does not need to be aware of whether `s.mu` is locked. Furthermore, the lock state of `s.mu` could change between the `preStart` call and its `handleStarted` callback.
```suggestion
func (s *service) preStart(c *runc.Container) (handleStarted func(*runc.Container, process.Process, muLocked bool), cleanup func()) {
```

Reply:
Nice advice. I'll fix it this way, if we really need to keep s.mu locked.